autofit/non_linear/parallel/sneaky.py

```
@@ -349,3 +350,171 @@ def __del__(self):
+            pass
+
+        self._processes = self.comm.size
+        
+        init_args = (
+            self.fitness_init, self.prior_transform_init,
+            self.fitness_args, self.fitness_kwargs,
+            self.prior_transform_args, self.prior_transform_kwargs
+        )
+        initializer(*init_args)
+
+    def check_if_mpi(self):
+
+        return self._processes > 1
+
+    def is_master(self):
+
+        is_mpi = self.check_if_mpi()
+        if is_mpi:
+            return_value = self.comm.rank == 0
+        else:
+            return_value = True
+        
+        return return_value
+
+    def wait(self):
+
+        is_mpi = self.check_if_mpi()
+        if is_mpi:
+            self.pool.wait()
+        else:
+            pass
+            warnings.warn(
+                "Cannot wait for pool to finish if not using MPI"
+            )
+
+    def __enter__(self):
+        """
+        Activate the mp / mpi pool
+        """
+
+        use_mpi = self.check_if_mpi()
+
+        if use_mpi:
+
+            from schwimmbad import MPIPool
+
+            if self.is_master():
+                logger.info("... using Schwimmbad MPIPool")
+            self.pool = MPIPool(use_dill=True)
+
+        else:
+            if self.is_master():
+                logger.info("... using multiprocessing")
+            self.pool = mp.Pool(
+                processes=self.processes,
+            )
+
+        return self
+
+    def map(
+            self, function: Callable,
+            iterable: Iterable
+    ):
+        """
+        Map a function over an iterable using the map method
+        of the initialized pool.
+
+        Parameters
+        ----------
+        function
+            A function to map
+        iterable
+            An iterable to map over
+
+        """
+        return self.pool.map(function, iterable)
+    
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        try:
+            self.pool.terminate()
+        except:  # noqa
+            pass
```

Either catch a more specific exception or log the exception. Silently catching all exceptions can lead to very difficult to find bugs.